pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,709 @@
+#![allow(clippy::unused_unit)]
+#![allow(clippy::upper_case_acronyms)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod benchmarking;
+pub mod weights;
+
+pub use pallet::*;
+
+type PoolId = u32;
+
+use frame_support::{
+	ensure,
+	sp_runtime::traits::{BlockNumberProvider, One, Zero},
+	transactional, PalletId,
+};
+use frame_system::ensure_signed;
+
+use sp_arithmetic::{
+	traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub},
+	FixedU128, Permill,
+};
+
+use orml_traits::MultiCurrency;
+
+use codec::{Decode, Encode};
+use frame_support::sp_runtime::{traits::AccountIdConversion, RuntimeDebug};
+use hydradx_traits::AMM;
+use scale_info::TypeInfo;
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct GlobalPool<Period, Balance, AssetId, AccountId, BlockNumber: AtLeast32BitUnsigned + Copy> {
+	id: PoolId,
+	owner: AccountId,
+	updated_at: Period,
+	total_shares: Balance,
+	accumulated_rps: Balance,
+	accumulated_rps_start: Balance,
+	reward_currency: AssetId,
+	accumulated_rewards: Balance,
+	paid_accumulated_rewards: Balance,
+	yield_per_period: Permill,
+	planned_yielding_periods: Period,
+	blocks_per_period: BlockNumber,
+	incentivized_token: AssetId,
+	max_reward_per_period: Balance,
+}
+
+impl<Period, Balance: std::default::Default, AssetId, AccountId, BlockNumber: AtLeast32BitUnsigned + Copy>
+	GlobalPool<Period, Balance, AssetId, AccountId, BlockNumber>
+{
+	fn new(
+		id: PoolId,
+		updated_at: Period,
+		reward_currency: AssetId,
+		yield_per_period: Permill,
+		planned_yielding_periods: Period,
+		blocks_per_period: BlockNumber,
+		owner: AccountId,
+		incentivized_token: AssetId,
+		max_reward_per_period: Balance,
+	) -> Self {
+		Self {
+			accumulated_rewards: Default::default(),
+			accumulated_rps: Default::default(),
+			accumulated_rps_start: Default::default(),
+			paid_accumulated_rewards: Default::default(),
+			total_shares: Default::default(),
+			id,
+			updated_at,
+			reward_currency,
+			yield_per_period,
+			planned_yielding_periods,
+			blocks_per_period,
+			owner,
+			incentivized_token,
+			max_reward_per_period,
+		}
+	}
+}
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct LiquidityPool<Period, Balance> {
+	id: PoolId,
+	updated_at: Period,
+	total_shares: Balance,
+	accumulated_rps: Balance,
+	loyalty_curve: Option<LoyaltyCurve>,
+	stake_in_global_pool: Balance,
+	multiplier: u32, //this is multiplier/weight in glboal pool
+}
+
+impl<Period, Balance: std::default::Default> LiquidityPool<Period, Balance> {
+	fn new(id: PoolId, updated_at: Period, loyalty_curve: Option<LoyaltyCurve>, multiplier: u32) -> Self {
+		Self {
+			accumulated_rps: Default::default(),
+			stake_in_global_pool: Default::default(),
+			total_shares: Default::default(),
+			id,
+			updated_at,
+			loyalty_curve,
+			multiplier,
+		}
+	}
+}
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct LoyaltyCurve {
+	b: FixedU128,
```

Comment:
Something more descriptive would not be possible instead of just 'b'?

Reply:
I saw it only as a `b`. Maybe @poliwop can suggest better name

Reply:
Ah yes, b is standard notation for "y-intercept" in the U.S.

In this case it's the initial reward, so if b = 0.20, the liquidity mining rewards start at 20% of max and then grow from there the longer the liquidity stays in the pool.

So something like "initial_reward_percentage" would be appropriate.